File: leetcode_32.cpp
class Solution {
public:
  char *strStr(const char *heystack, const char *needle) {
    if (!*needle) return heystack;

    const char *p1;
    const char *p2;
    const char *p1_advance = heystack;

    for (p2 = &needle[1]; *p2; p2++) {
      p1_advance++;
    }

    for (p1 = heystack; *p1_advance; p1_advance++) {
      const char *p1_old = p1;
      p2 = needle;
      while (*p1 && *p2 && *p1 == *p2) {
        p1++;
        p2++;
      }
      if (!*p2) return p1_old;
      p1 = p1_old + 1;
    }

    return nullptr;
  }
}

File: leetcode_32.py
class Solution:
    # @param haystack, a string
    # @param needle, a string
    # @return a string or None
    def strStr(self, haystack, needle):
        h_len = len(haystack)
        n_len = len(needle)
        x = h_len - n_len + 1
        if (x < 1):
            return None
        result = True
        pivot = 0
        for i in range(x):
            result = True
            for j in range(n_len):
                if needle[j] != haystack[i + j]:
                    result = False
                    break
            if result:
                pivot = i
                break
        if result:
            return haystack[pivot:]
        else:
            return None


s = Solution()
print s.strStr('hello', 'e')
